### PR TITLE
Add more c apis correponding to python c apis.

### PR DESCRIPTION
**Pull request by SanftMonster**
Hello everyone,

I would like to submit this PR that adds some C APIs corresponding to those in `tensorflow/c/python_api.h`. The background behind this is that I am a developer of [Tensorflow.NET](http://tensorflow.net/), which is a .NET binding of Tensorflow and is still actively maintained. When adding new features, the best approach for us is to write C# code by referring to Python code. However, sometimes Python uses C APIs defined in `tensorflow/c/python_api.h` but these APIs are not included in common externed C APIs, making it difficult for us to add the particular feature.

To address this issue, I have added APIs in `tensorflow/c/python_api.h` to `tensorflow/c/c_api.h`, as these APIs are commonly required when adding features, with Tensorflow Python being the only reliable reference. I have tested this PR by compiling this branch on both Windows and Linux and it works. For example, the latest versions of [Tensorflow.Redist 2.11](https://www.nuget.org/packages/SciSharp.TensorFlow.Redist) and [Tensorflow.Redist-Windows-GPU 2.10](https://www.nuget.org/packages/SciSharp.TensorFlow.Redist-Windows-GPU) both work smoothly. (We also successfully compiled the Linux GPU version, but it has not been released due to package size limits of NuGet)

If adding too many APIs once is not encouraged, then I kindly request to add `TF_SetAttr`, `TF_UpdateEdge`, `TF_GetHandleShapeAndType`, and `TF_SetHandleShapeAndType`.

Thank you very much for your consideration. :)

### CODE REVIEW

**Comment by AsakusaRinne** [comment]
@sagunb Sorry for the interruption but I met some problem of ci while I'm fresh to tensorflow. The ci `AMD ROCm -- Community CI Build` said the bazel version does not match. However I didn't change any bazel version setting in this PR. Could you please help to see what I should do resolve this error?

File: tensorflow/c/c_api.h [hunk 1]
```diff
@@ -1577,6 +1578,70 @@ TF_CAPI_EXPORT extern void TF_RegisterLogListener(
...
+
+// Sets the experimental_type` field in the node_def Protocol Buffer.
+TF_CAPI_EXPORT extern void TF_SetFullType(TF_Graph* graph, TF_Operation* op,
+                                           const tensorflow::FullTypeDef& full_type);
+
+TF_CAPI_EXPORT extern void TF_SetRequestedDevice(TF_Graph* graph,
+                                                  TF_Operation* op,
+                                                  const char* device);
+
+TF_CAPI_EXPORT extern void TF_RemoveAllControlInputs(TF_Graph* graph, TF_Operation* op);
```
Comment (rohan100jain) [inline tensorflow/c/c_api.h:1607]: Please add docstrings for both these APIs

**Review by rohan100jain — APPROVED** [review]

File: tensorflow/c/c_api.h (hunk 1 above)
Comment (AsakusaRinne) [reply tensorflow/c/c_api.h:1609]: Thank you for your reminder. I've added docstrings to them.

**Review by sagunb — APPROVED** [review]

**Comment by AsakusaRinne** [comment]
Could you please help with the ci error? The result showed `__main__.UserInputError: Invalid CLANG_COMPILER_PATH setting was provided 10 times in a row. Assuming to be a scripting mistake.`, which seems to be a clang config error. However I didn't modify related settings in this PR.

**Review by sagunb — APPROVED** [review]

**Comment by sagunb** [comment]
Hi @AsakusaRinne I've ported over this PR to the internal tool and am running it against the internal checks to see what all might be failing. I will update you if anything pops up. Nothing new should be required from you at this point.

**Pull request merged**